File: static/src/js/cafm_dashboard.js
import { registry } from "@web/core/registry";
import { Component, onWillStart, useState, onMounted } from "@odoo/owl";
import { useService } from "@web/core/utils/hooks";
import { loadJS } from "@web/core/assets";

class FacilityDashboard extends Component {
    setup() {
        this.orm = useService("orm");
        this.action = useService("action");

        this.state = useState({
            data: {
                // Assets
                total_assets: 0,
                assets_active: 0,
                assets_maintenance: 0,
                assets_out_of_service: 0,
                assets_retired: 0,
                assets_by_company: "[]",
                assets_by_category: "[]",

                // Maintenance
                total_maintenance: 0,
                maintenance_scheduled: 0,
                maintenance_in_progress: 0,
                maintenance_pending_parts: 0,
                maintenance_completed: 0,
                maintenance_cancelled: 0,
                maintenance_overdue: 0,

                maintenance_preventive: 0,
                maintenance_corrective: 0,
                maintenance_predictive: 0,

                // Helpdesk
                total_helpdesk: 0,
                helpdesk_new: 0,
                helpdesk_open: 0,
                helpdesk_in_progress: 0,
                helpdesk_pending: 0,
                helpdesk_resolved: 0,
                helpdesk_closed: 0,
                helpdesk_unassigned: 0,

                helpdesk_urgent: 0,
                helpdesk_high: 0,
                helpdesk_normal: 0,
                helpdesk_low: 0,

                // Teams
                total_teams: 0,
                total_team_members: 0,
                teams_available: 0,
                teams_busy: 0,
                team_workload: "[]",

                // Trends
                monthly_maintenance_data: "[]",
                monthly_helpdesk_data: "[]",
            },
            loading: true,
        });

        onWillStart(async () => {
            await loadJS("/web/static/lib/Chart/Chart.js");
            await this.loadDashboardData();
        });

        onMounted(() => {
            if (!this.state.loading) {
                this.renderCharts();
            }
        });
    }

    async loadDashboardData() {
        try {
            const ids = await this.orm.call("asset.dashboard", "search", [[]]);
            let id = ids.length ? ids[0] : await this.orm.call("asset.dashboard", "create", [{}]);

            const data = await this.orm.call("asset.dashboard", "read", [
                [id],
                Object.keys(this.state.data),
            ]);

            if (data && data.length) {
                this.state.data = { ...this.state.data, ...data[0] };
            }

            this.state.loading = false;
            setTimeout(() => this.renderCharts(), 50);

        } catch (error) {
            console.error("Dashboard Load Error", error);
            this.state.loading = false;
        }
    }

    renderCharts() {
        if (this.state.loading) return;

        try {
            this.renderAssetsCompanyChart();
            this.renderAssetsCategoryChart();
            this.renderMaintenanceStatusChart();
            this.renderMaintenanceTypeChart();
            this.renderMaintenanceTypeBarChart();
            this.renderHelpdeskStatusChart();
            this.renderHelpdeskPriorityChart();
            this.renderTeamWorkloadChart();
            this.renderMonthlyMaintenanceChart();
            this.renderMonthlyHelpdeskChart();
        } catch (error) {
            console.error("Chart Render Error:", error);
        }
    }

    // --------------------------------------------------------------
    //  ASSETS BY COMPANY (NEW REPLACEMENT)
    // --------------------------------------------------------------
    renderAssetsCompanyChart() {
        const canvas = document.getElementById("assetsCompanyChart");
        if (!canvas) return;

        const existing = Chart.getChart(canvas);
        if (existing) existing.destroy();

        let items = [];
        try {
            items = JSON.parse(this.state.data.assets_by_company || "[]");
        } catch (e) {}

        new Chart(canvas, {
            type: "bar",
            data: {
                labels: items.map((x) => x.company),
                datasets: [{
                    label: "Assets",
                    data: items.map((x) => x.count),
                    backgroundColor: "#0d6efd",
                }],
            },
            options: {
                responsive: true,
                maintainAspectRatio: false,
                plugins: { legend: { display: false } },
                scales: { y: { beginAtZero: true } },
            },
        });
    }



    // --------------------------------------------------------------
    //  ASSETS BY CATEGORY
    // --------------------------------------------------------------
    renderAssetsCategoryChart() {
        const canvas = document.getElementById("assetsCategoryChart");
        if (!canvas) return;

        const existing = Chart.getChart(canvas);
        if (existing) existing.destroy();

        let items = [];
        try {
            items = JSON.parse(this.state.data.assets_by_category || "[]");
        } catch (e) {}

        new Chart(canvas, {
            type: "bar",
            data: {
                labels: items.map((d) => d.name),
                datasets: [{
                    label: "Assets",
                    data: items.map((d) => d.count),
                    backgroundColor: "#198754",
                }],
            },
            options: {
                responsive: true,
                maintainAspectRatio: false,
                scales: { y: { beginAtZero: true } },
                plugins: { legend: { display: false } },
            },
        });
    }

    // --------------------------------------------------------------
    //  MAINTENANCE STATUS
    // --------------------------------------------------------------
    renderMaintenanceStatusChart() {
        const canvas = document.getElementById("maintenanceStatusChart");
        if (!canvas) return;

        const ex = Chart.getChart(canvas);
        if (ex) ex.destroy();

        new Chart(canvas, {
            type: "doughnut",
            data: {
                labels: ["Scheduled", "In Progress", "Pending Parts", "Completed", "Cancelled"],
                datasets: [{
                    data: [
                        this.state.data.maintenance_scheduled,
                        this.state.data.maintenance_in_progress,
                        this.state.data.maintenance_pending_parts,
                        this.state.data.maintenance_completed,
                        this.state.data.maintenance_cancelled,
                    ],
                    backgroundColor: [
                        "#6c757d",
                        "#0d6efd",
                        "#ffc107",
                        "#198754",
                        "#dc3545"
                    ],
                }],
            },
            options: {
                responsive: true,
                plugins: { legend: { position: "bottom" } },
            },
        });
    }

    // --------------------------------------------------------------
    //  MAINTENANCE TYPE
    // --------------------------------------------------------------
    renderMaintenanceTypeChart() {
        const canvas = document.getElementById("maintenanceTypeChart");
        if (!canvas) return;

        const ex = Chart.getChart(canvas);
        if (ex) ex.destroy();

        new Chart(canvas, {
            type: "pie",
            data: {
                labels: ["Preventive", "Corrective", "Predictive"],
                datasets: [{
                    data: [
                        this.state.data.maintenance_preventive,
                        this.state.data.maintenance_corrective,
                        this.state.data.maintenance_predictive,
                    ],
                    backgroundColor: ["#198754", "#ffc107", "#0d6efd"],
                }],
            },
            options: {
                responsive: true,
                plugins: { legend: { position: "bottom" } },
            },
        });
    }

    renderMaintenanceTypeBarChart() {
    const ctx = document.getElementById("maintenanceTypeBarChart");
    if (!ctx) return;

    const existingChart = Chart.getChart(ctx);
    if (existingChart) existingChart.destroy();

    new Chart(ctx, {
        type: "bar",
        data: {
            labels: ["Preventive", "Corrective"],
            datasets: [{
                label: "Tasks",
                data: [
                    this.state.data.maintenance_preventive || 0,
                    this.state.data.maintenance_corrective || 0,
                ],
                backgroundColor: ["#198754", "#dc3545"],
            }],
        },
        options: {
            responsive: true,
            maintainAspectRatio: false,
            scales: {
                y: { beginAtZero: true }
            },
            plugins: { legend: { display: false } },
        },
    });
    }

    // --------------------------------------------------------------
    //  HELPDESK STATUS
    // --------------------------------------------------------------
    renderHelpdeskStatusChart() {
        const canvas = document.getElementById("helpdeskStatusChart");
        if (!canvas) return;

        const ex = Chart.getChart(canvas);
        if (ex) ex.destroy();

        new Chart(canvas, {
            type: "bar",
            data: {
                labels: ["New", "Open", "In Progress", "Pending", "Resolved", "Closed"],
                datasets: [{
                    label: "Tickets",
                    data: [
                        this.state.data.helpdesk_new,
                        this.state.data.helpdesk_open,
                        this.state.data.helpdesk_in_progress,
                        this.state.data.helpdesk_pending,
                        this.state.data.helpdesk_resolved,
                        this.state.data.helpdesk_closed,
                    ],
                    backgroundColor: [
                        "#6c757d",
                        "#0d6efd",
                        "#ffc107",
                        "#fd7e14",
                        "#198754",
                        "#20c997",
                    ],
                }],
            },
            options: {
                responsive: true,
                scales: { y: { beginAtZero: true } },
                plugins: { legend: { display: false } },
            },
        });
    }

    // --------------------------------------------------------------
    //  HELPDESK PRIORITY
    // --------------------------------------------------------------
    renderHelpdeskPriorityChart() {
        const canvas = document.getElementById("helpdeskPriorityChart");
        if (!canvas) return;

        const ex = Chart.getChart(canvas);
        if (ex) ex.destroy();

        new Chart(canvas, {
            type: "doughnut",
            data: {
                labels: ["Urgent", "High", "Normal", "Low"],
                datasets: [{
                    data: [
                        this.state.data.helpdesk_urgent,
                        this.state.data.helpdesk_high,
                        this.state.data.helpdesk_normal,
                        this.state.data.helpdesk_low,
                    ],
                    backgroundColor: ["#dc3545", "#fd7e14", "#ffc107", "#198754"],
                }],
            },
            options: {
                responsive: true,
                plugins: { legend: { position: "bottom" } },
            },
        });
    }

    // --------------------------------------------------------------
    //  TEAM WORKLOAD
    // --------------------------------------------------------------
    renderTeamWorkloadChart() {
        const canvas = document.getElementById("teamWorkloadChart");
        if (!canvas) return;

        const existing = Chart.getChart(canvas);
        if (existing) existing.destroy();

        let items = [];
        try {
            items = JSON.parse(this.state.data.team_workload || "[]");
        } catch (e) {}

        new Chart(canvas, {
            type: "bar",
            data: {
                labels: items.map((d) => d.team),
                datasets: [{
                    label: "Active Tasks",
                    data: items.map((d) => d.tasks),
                    backgroundColor: "#0d6efd",
                }],
            },
            options: {
                responsive: true,
                scales: { y: { beginAtZero: true } },
                plugins: { legend: { display: false } },
            },
        });
    }

    // --------------------------------------------------------------
    //  MONTHLY MAINTENANCE TREND
    // --------------------------------------------------------------
    renderMonthlyMaintenanceChart() {
        const canvas = document.getElementById("monthlyMaintenanceChart");
        if (!canvas) return;

        const ex = Chart.getChart(canvas);
        if (ex) ex.destroy();

        let items = [];
        try {
            items = JSON.parse(this.state.data.monthly_maintenance_data || "[]");
        } catch (e) {}

        new Chart(canvas, {
            type: "line",
            data: {
                labels: items.map((d) => d.month),
                datasets: [
                    {
                        label: "Total Tasks",
                        data: items.map((d) => d.total),
                        borderColor: "#0d6efd",
                        backgroundColor: "rgba(13,110,253,0.2)",
                        tension: 0.3,
                        fill: true,
                    },
                    {
                        label: "Completed",
                        data: items.map((d) => d.completed),
                        borderColor: "#198754",
                        backgroundColor: "rgba(25,135,84,0.2)",
                        tension: 0.3,
                        fill: true,
                    },
                ],
            },
            options: {
                responsive: true,
                plugins: { legend: { position: "top" } },
            },
        });
    }

    // --------------------------------------------------------------
    //  MONTHLY HELPDESK TREND
    // --------------------------------------------------------------
    renderMonthlyHelpdeskChart() {
        const canvas = document.getElementById("monthlyHelpdeskChart");
        if (!canvas) return;

        const ex = Chart.getChart(canvas);
        if (ex) ex.destroy();

        let items = [];
        try {
            items = JSON.parse(this.state.data.monthly_helpdesk_data || "[]");
        } catch (e) {}

        new Chart(canvas, {
            type: "line",
            data: {
                labels: items.map((d) => d.month),
                datasets: [
                    {
                        label: "Total Tickets",
                        data: items.map((d) => d.total),
                        borderColor: "#0d6efd",
                        backgroundColor: "rgba(13,110,253,0.2)",
                        tension: 0.3,
                        fill: true,
                    },
                    {
                        label: "Resolved",
                        data: items.map((d) => d.resolved),
                        borderColor: "#198754",
                        backgroundColor: "rgba(25,135,84,0.2)",
                        tension: 0.3,
                        fill: true,
                    },
                ],
            },
            options: {
                responsive: true,
                plugins: { legend: { position: "top" } },
            },
        });
    }

    // Open list view
    async openView(model, name, domain = []) {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: name,
            res_model: model,
            views: [[false, "list"], [false, "form"]],
            domain,
        });
    }
}

FacilityDashboard.template = "facility_management.Dashboard";
registry.category("actions").add("facility_dashboard", FacilityDashboard);

File: models/cafm_dashboard.py
from odoo import api, fields, models
from datetime import datetime, timedelta
from dateutil.relativedelta import relativedelta
import json

class FacilityDashboard(models.Model):
    _name = 'asset.dashboard'
    _description = 'Facility Management Dashboard'

    name = fields.Char(default="Facility Dashboard")

    # ASSET STATISTICS
    total_assets = fields.Integer(compute="_compute_asset_stats")
    assets_active = fields.Integer(compute="_compute_asset_stats")
    assets_maintenance = fields.Integer(compute="_compute_asset_stats")
    assets_out_of_service = fields.Integer(compute="_compute_asset_stats")
    assets_retired = fields.Integer(compute="_compute_asset_stats")

    assets_by_company = fields.Text(compute="_compute_assets_by_company")
    assets_by_category = fields.Text(compute="_compute_asset_category_stats")

    # MAINTENANCE STATS
    total_maintenance = fields.Integer(compute="_compute_maintenance_stats")
    maintenance_scheduled = fields.Integer(compute="_compute_maintenance_stats")
    maintenance_in_progress = fields.Integer(compute="_compute_maintenance_stats")
    maintenance_pending_parts = fields.Integer(compute="_compute_maintenance_stats")
    maintenance_completed = fields.Integer(compute="_compute_maintenance_stats")
    maintenance_cancelled = fields.Integer(compute="_compute_maintenance_stats")
    maintenance_overdue = fields.Integer(compute="_compute_maintenance_stats")

    maintenance_preventive = fields.Integer(compute="_compute_maintenance_type")
    maintenance_corrective = fields.Integer(compute="_compute_maintenance_type")
    maintenance_predictive = fields.Integer(compute="_compute_maintenance_type")

    # HELPDESK
    total_helpdesk = fields.Integer(compute="_compute_helpdesk_stats")
    helpdesk_new = fields.Integer(compute="_compute_helpdesk_stats")
    helpdesk_open = fields.Integer(compute="_compute_helpdesk_stats")
    helpdesk_in_progress = fields.Integer(compute="_compute_helpdesk_stats")
    helpdesk_pending = fields.Integer(compute="_compute_helpdesk_stats")
    helpdesk_resolved = fields.Integer(compute="_compute_helpdesk_stats")
    helpdesk_closed = fields.Integer(compute="_compute_helpdesk_stats")
    helpdesk_unassigned = fields.Integer(compute="_compute_helpdesk_stats")

    helpdesk_urgent = fields.Integer(compute="_compute_helpdesk_priority")
    helpdesk_high = fields.Integer(compute="_compute_helpdesk_priority")
    helpdesk_normal = fields.Integer(compute="_compute_helpdesk_priority")
    helpdesk_low = fields.Integer(compute="_compute_helpdesk_priority")

    # TEAMS
    total_teams = fields.Integer(compute="_compute_team_stats")
    total_team_members = fields.Integer(compute="_compute_team_stats")
    teams_available = fields.Integer(compute="_compute_team_stats")
    teams_busy = fields.Integer(compute="_compute_team_stats")
    team_workload = fields.Text(compute="_compute_team_workload")

    # TRENDS
    monthly_maintenance_data = fields.Text(compute="_compute_monthly_maintenance")
    monthly_helpdesk_data = fields.Text(compute="_compute_monthly_helpdesk")


    # ----------------------------
    # ASSET COMPUTATION
    # ----------------------------
    def _compute_asset_stats(self):
        Asset = self.env['account.asset']
        company = self.env.company.id
        domain = [
            ('company_id', '=', company),
        ]

        for rec in self:
            rec.total_assets = Asset.search_count(domain)
            rec.assets_active = Asset.search_count(domain + [('state', '=', 'active')])
            rec.assets_maintenance = Asset.search_count(domain + [('state', '=', 'maintenance')])
            rec.assets_out_of_service = Asset.search_count(domain + [('state', '=', 'out_of_service')])
            rec.assets_retired = Asset.search_count(domain + [('state', '=', 'retired')])

    def _compute_assets_by_company(self):
        for rec in self:
            query = """
                SELECT customer_id, COUNT(id)
                FROM account_asset
                WHERE customer_id IS NOT NULL
                GROUP BY customer_id;
            """

            self.env.cr.execute(query)
            rows = self.env.cr.fetchall()

            result = []

            for customer_id, count in rows:
                customer_name = self.env['res.partner'].browse(customer_id).name
                result.append({
                    'company': customer_name,
                    'count': count,
                })

            rec.assets_by_company = json.dumps(result)

    def _compute_asset_category_stats(self):
        for rec in self:
            field = self.env['account.asset']._fields['asset_category']
            categories = dict(field.selection)
            result = []

            for key, label in categories.items():
                count = self.env['account.asset'].search_count([('asset_category', '=', key)])
                result.append({"name": label, "count": count})

            rec.assets_by_category = json.dumps(result)


    # ----------------------------
    # MAINTENANCE
    # ----------------------------
    def _compute_maintenance_stats(self):
        Task = self.env['asset.maintenance.task']
        for rec in self:
            rec.total_maintenance = Task.search_count([])

            rec.maintenance_scheduled = Task.search_count([('state', '=', 'draft')])
            rec.maintenance_in_progress = Task.search_count([('state', '=', 'in_progress')])
            rec.maintenance_pending_parts = 0  # no such state in your model
            rec.maintenance_completed = Task.search_count([('state', '=', 'done')])
            rec.maintenance_cancelled = Task.search_count([('state', '=', 'cancel')])

            today = fields.Date.today()
            rec.maintenance_overdue = Task.search_count([
                ('scheduled_date', '<', today),
                ('state', 'not in', ['done', 'cancel'])
            ])

    def _compute_maintenance_type(self):
        Task = self.env['asset.maintenance.task']
        for rec in self:
            rec.maintenance_preventive = Task.search_count([('maintenance_type', '=', 'preventive')])
            rec.maintenance_corrective = Task.search_count([('maintenance_type', '=', 'corrective')])
            rec.maintenance_predictive = Task.search_count([('maintenance_type', '=', 'predictive')])


    # ----------------------------
    # HELPDESK
    # ----------------------------
    def _compute_helpdesk_stats(self):
        Ticket = self.env['helpdesk.ticket']
        for rec in self:
            rec.total_helpdesk = Ticket.search_count([])
            rec.helpdesk_new = Ticket.search_count([('stage_id.name', '=', 'New')])
            rec.helpdesk_open = Ticket.search_count([('stage_id.name', '=', 'Open')])
            rec.helpdesk_in_progress = Ticket.search_count([('stage_id.name', '=', 'In Progress')])
            rec.helpdesk_pending = Ticket.search_count([('stage_id.name', '=', 'On Hold')])
            rec.helpdesk_resolved = Ticket.search_count([('stage_id.name', '=', 'Solved')])
            rec.helpdesk_closed = Ticket.search_count([('stage_id.name', '=', 'Closed')])
            rec.helpdesk_unassigned = Ticket.search_count([('user_id', '=', False)])


    def _compute_helpdesk_priority(self):
        Ticket = self.env['helpdesk.ticket']
        for rec in self:
            rec.helpdesk_urgent = Ticket.search_count([('priority', '=', '3')])
            rec.helpdesk_high = Ticket.search_count([('priority', '=', '2')])
            rec.helpdesk_normal = Ticket.search_count([('priority', '=', '1')])
            rec.helpdesk_low = Ticket.search_count([('priority', '=', '0')])


    # ----------------------------
    # TEAM COMPUTATION
    # ----------------------------
    def _compute_team_stats(self):
        Team = self.env['asset.maintenance.team']
        Task = self.env['asset.maintenance.task']

        for rec in self:
            teams = Team.search([])
            rec.total_teams = len(teams)
            rec.total_team_members = sum(len(team.member_ids) for team in teams)

            busy = Task.search([('state', '=', 'in_progress')]).mapped('maintenance_team_id')
            rec.teams_busy = len(busy)
            rec.teams_available = rec.total_teams - rec.teams_busy


    def _compute_team_workload(self):
        Task = self.env['asset.maintenance.task']
        Team = self.env['asset.maintenance.team']

        for rec in self:
            result = []

            for team in Team.search([]):
                task_count = Task.search_count([
                    ('maintenance_team_id', '=', team.id),
                    ('state', 'in', ['draft', 'in_progress'])
                ])
                result.append({
                    "team": team.name,
                    "tasks": task_count,
                    "members": len(team.member_ids)
                })

            rec.team_workload = json.dumps(result)


    # ----------------------------
    # MONTHLY TRENDS
    # ----------------------------
    def _compute_monthly_maintenance(self):
        Task = self.env['asset.maintenance.task']

        for rec in self:
            data = []
            today = datetime.today()

            for i in range(11, -1, -1):
                month = today - relativedelta(months=i)
                start = month.replace(day=1)
                end = (start + relativedelta(months=1)) - timedelta(seconds=1)

                total = Task.search_count([
                    ('create_date', '>=', start),
                    ('create_date', '<=', end),
                ])

                completed = Task.search_count([
                    ('create_date', '>=', start),
                    ('create_date', '<=', end),
                    ('state', '=', 'done'),
                ])

                data.append({
                    "month": month.strftime("%b %Y"),
                    "total": total,
                    "completed": completed,
                })

            rec.monthly_maintenance_data = json.dumps(data)


    def _compute_monthly_helpdesk(self):
        Ticket = self.env['helpdesk.ticket']

        for rec in self:
            data = []
            today = datetime.today()

            for i in range(11, -1, -1):
                month = today - relativedelta(months=i)
                start = month.replace(day=1)
                end = (start + relativedelta(months=1)) - timedelta(seconds=1)

                total = Ticket.search_count([
                    ('create_date', '>=', start),
                    ('create_date', '<=', end),
                ])

                resolved = Ticket.search_count([
                    ('stage_id.name', '=', 'Solved'),
                    ('create_date', '>=', start),
                    ('create_date', '<=', end),
                ])

                data.append({
                    "month": month.strftime("%b %Y"),
                    "total": total,
                    "resolved": resolved,
                })

            rec.monthly_helpdesk_data = json.dumps(data)
